ipvs.go

```diff
@@ -199,23 +199,37 @@ func parseIPVSBackendStatus(file io.Reader) ([]IPVSBackendStatus, error) {
 	return status, nil
 }
 
-func parseIPPort(s string) (net.IP, uint16, error) {
-	tmp := strings.SplitN(s, ":", 2)
-
-	if len(tmp) != 2 {
-		return nil, 0, fmt.Errorf("invalid IP:Port: %s", s)
-	}
+func parseIPPort(ipPort string) (net.IP, uint16, error) {
+	var (
+		portString string
+		ip         net.IP
+		err        error
+	)
 
-	if len(tmp[0]) != 8 && len(tmp[0]) != 32 {
-		return nil, 0, fmt.Errorf("invalid IP: %s", tmp[0])
+	if len(ipPort) == 13 || len(ipPort) == 37 {
+		tmp := strings.SplitN(ipPort, ":", 2)
+		if len(tmp) != 2 {
+			return nil, 0, fmt.Errorf("invalid IP:Port: %s", ipPort)
+		}
+		portString = tmp[1]
+		ip, err = hex.DecodeString(tmp[0])
+		if err != nil {
+			return nil, 0, err
+		}
+	} else if len(ipPort) == 46 && strings.HasPrefix(ipPort, "[") {
+		portString = ipPort[42:]
+		ip = net.ParseIP(ipPort[1:40])
+		if ip == nil {
+			return nil, 0, fmt.Errorf("invalid IPv6 address: %s", ipPort[1:40])
+		}
+	} else {
+		return nil, 0, fmt.Errorf("invalid IP:Port: %s", ipPort)
 	}
 
-	ip, err := hex.DecodeString(tmp[0])
-	if err != nil {
-		return nil, 0, err
+	if len(portString) != 4 {
+		return nil, 0, fmt.Errorf("invalid port: %s", portString)
```

Comment:
I don't like the usage of "invalid" here. More likely than being an invalid port, we'll be presented a different, unexpected format. So I'd change the wording in most of the error messages to `unexpected`.